### PR TITLE
Refactor Witch class

### PR DESCRIPTION
**Pull request by neihousaigaai**

### CODE REVIEW

File: game/__init__.py [hunk 1]
```diff
@@ -140,6 +140,10 @@ def set_mode(self, mode_id, status):
                 self.new_moon_mode.turn_on()
             else:
                 self.new_moon_mode.turn_off()
+
+        elif mode_str == "witch_can_kill":
+            roles.Witch.set_can_kill(status=="on")
+
```
Comment (nitro2) [inline game/__init__.py:146]: I think this setting only affects "before" the game starts

File: game/__init__.py [hunk 2]
```diff
@@ -806,18 +812,22 @@ async def do_new_nighttime_phase(self):
...
             await self.interface.send_embed_to_channel(embed_data, config.WEREWOLF_CHANNEL)
             # Send alive player list to all skilled characters (guard, seer, etc.)
-            if self.modes.get("witch_can_kill"):
-                await asyncio.gather(*[player.on_action(embed_data) for player in self.get_alive_players()])
-            else:
-                await asyncio.gather(*[player.on_action(embed_data) for player in self.get_alive_players() if not isinstance(player, roles.Witch)])
+            await asyncio.gather(*[
+                player.on_action(embed_data) for player in self.get_alive_players()
+                if not isinstance(player, roles.Witch) or (roles.Witch.is_can_kill() and player.get_curse_power())
+            ])
```
Comment (nitro2) [inline game/__init__.py:822]: Should we move these to Role.on_night_phase()?

File: game/__init__.py [hunk 3]
```diff
@@ -919,22 +929,23 @@ async def seer_do_end_nighttime_phase(self, author):
...
 
-        curse_target_id = author.get_curse_target()
-        if author.get_curse_power() > 0 and curse_target_id:
-            author.on_use_curse_power()
-            self.night_pending_kill_list.append(curse_target_id)
+        if roles.Witch.is_can_kill():
+            curse_target_id = author.get_curse_target()
+            if author.get_curse_power() > 0 and curse_target_id:
+                author.on_use_curse_power()
+                self.night_pending_kill_list.append(curse_target_id)
```
Comment (nitro2) [inline game/__init__.py:944]: How about if Witch can kill, we provide `curse_power = 1` ? Not sure which one is clearer

File: game/__init__.py [hunk 4]
```diff
@@ -1168,6 +1179,7 @@ async def reborn(self, author, target):
     @command_verify_author(roles.Witch)
     @command_verify_phase(const.GamePhase.NIGHT)
     async def curse(self, author, target):
+        roles.Witch.set_can_kill(self.modes.get("witch_can_kill", False))
```
Comment (nitro2) [inline game/__init__.py:1182]: this should check `roles.Witch.is_can_kill()` instead of set?

Comment (neihousaigaai) [reply game/__init__.py:1182]: Will remove it.

File: game/__init__.py (hunk 3 above)
Comment (neihousaigaai) [reply game/__init__.py:944]: It depends on game specs, now `curse` effect is killing people. Dunno if we will apply another effects for cursed target or not.

File: game/__init__.py (hunk 2 above)
Comment (neihousaigaai) [reply game/__init__.py:822]: Will refactor it in next PR.

**Review by nitro2 — APPROVED** [review]

File: game/__init__.py (hunk 1 above)
Comment (neihousaigaai) [reply game/__init__.py:146]: Will remove in next PR.

**Pull request merged**